dashboard/tasks.py

```diff
@@ -2,36 +2,49 @@
 Periodic task that updates user data.
 """
 import logging
-import uuid
-from datetime import timedelta
 
 from celery import group
-from django.conf import settings
-from django.contrib.auth.models import User
 from django.core.cache import caches
-from django.db.models import Q
-from edx_api.client import EdxApi
-from social_django.models import UserSocialAuth
 
-from backends import utils
-from dashboard.models import UserCacheRefreshTime
-from dashboard.api_edx_cache import CachedEdxDataApi
-from micromasters.celery import app
-from micromasters.utils import (
-    chunks,
-    now_in_utc,
+from dashboard.api import (
+    calculate_users_to_refresh_in_bulk,
+    refresh_user_data,
 )
-from profiles.api import get_social_auth
+from micromasters.celery import app
+from micromasters.utils import chunks
 
 
 log = logging.getLogger(__name__)
 cache_redis = caches['redis']
 
 
-MAX_HRS_MAIN_TASK = 6
 PARALLEL_RATE_LIMIT = '10/m'
 LOCK_EXPIRE = 60 * 60 * 2  # Lock expires in 2 hrs
-lock_id = '{0}-lock'.format(uuid.uuid4())
+LOCK_ID = 'batch_update_user_data_lock'
+
+
+def _acquire_lock():
+    """
+    create lock by adding a key to storage.
+    """
+    # lock will expire if 2 hrs are pass and task is not complete.
+    return cache_redis.add(LOCK_ID, 'true', LOCK_EXPIRE)
+
+
+def _release_lock():
+    """
+    remove lock by deleting key from storage.
+    """
+    return cache_redis.delete(LOCK_ID)
```

Comment:
To have a truly safe distributed lock you need to set the key to a unique value and then only release it if they value matches the value you set it as (so you don't release a lock created by another process). Since we're only holding the lock for 2 hours but only run the task every 6 we should be ok, but it might be worth adding a note of this limitation in case we change how often we run it down the road.

Reply:
Ok I'll add a comment

Reply:
Actually I'm thinking we should just get rid of the subtasks altogether and check the lock similar to how it's done for `PercolateQueryMembership`.

Reply:
Changed my mind again, I'll just add the comment to this PR and then work on the refactor in the next PR, which will also include throttling changes